common/component/redis/redis.go

```diff
@@ -187,11 +210,135 @@ func ParseClientFromProperties(properties map[string]string, componentType metad
 			newClientFunc = newV9FailoverClient
 		}
 	}
-	c, err = newClientFunc(settings)
-	if err != nil {
-		return nil, nil, fmt.Errorf("redis client configuration error: %w", err)
+	c, newClientErr = newClientFunc(&settings)
+	if newClientErr != nil {
+		return nil, nil, fmt.Errorf("redis client configuration error: %w", newClientErr)
+	}
+
+	// start the token refresh goroutine
+
+	if settings.UseEntraID {
+		StartEntraIDTokenRefreshBackgroundRoutine(c, settings.Username, *tokenExpires, tokenCredential, logger)
+	}
+	return c, &settings, nil
+}
+
+func StartEntraIDTokenRefreshBackgroundRoutine(client RedisClient, username string, nextExpiration time.Time, cred *azcore.TokenCredential, logger *kitlogger.Logger) {
+	go func(cred *azcore.TokenCredential, username string, logger *kitlogger.Logger) {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		backoffConfig := kitretry.DefaultConfig()
+		backoffConfig.MaxRetries = 3
+		backoffConfig.Policy = kitretry.PolicyExponential
+
+		var backoffManager backoff.BackOff
+		refreshGracePeriod := 2 * time.Minute
+		tokenRefreshDuration := time.Until(nextExpiration.Add(-refreshGracePeriod))
+
+		(*logger).Debugf("redis client: starting entraID token refresh loop")
+
+		for {
+			(*logger).Debugf("redis client: next entraID token refresh: %v", tokenRefreshDuration)
+			select {
+			case <-ctx.Done():
```

Comment:
Because `ctx` is derived by `context.Background()`, the context is never canceled. So if the component is stopped, we don't stop refreshing tokens. Additionally, if the component is reloaded, we spin a new goroutine for each time